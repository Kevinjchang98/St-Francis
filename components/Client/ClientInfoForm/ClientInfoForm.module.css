.container {
    display: flex;
    flex-direction: column;
    width: 80vw;
    max-width: 1000px;
}

.titleContainer {
    display: flex;
    flex-direction: row;
    align-items: center;
    border-bottom: 0.5px solid var(--gray-3);
}

.titleContainer h1 {
    flex-grow: 1;
}

.formRows {
    display: flex;
    flex-wrap: wrap;
}

.formRowItem {
    flex-grow: 1;
    flex-shrink: 0;
    flex-basis: calc(25% - 10px);
    margin: 5px;
    align-items: center;
    justify-content: center;
}

.container input,
.container label {
    display: block;
}

.container textarea {
    width: calc(100% - 10px);
    margin: 5px;
}

.saveButtons {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin-top: 25px;
    margin-left: -25px;
}

.saveButtons .backButton {
    width: 200px;
    margin-left: 25px;
}

.saveButtons .saveButton {
    width: 200px;
    margin-left: 25px;
}

.saveButtons span {
    flex-grow: 1;
}
