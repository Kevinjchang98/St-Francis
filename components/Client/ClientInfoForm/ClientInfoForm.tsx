'use client';
import { Client } from '@/models/index';
import { DocFilter } from '@/utils/index';
import Link from 'next/link';
import { useState } from 'react';
import styles from './ClientInfoForm.module.css';
interface ClientInfoFormProps {
    id?: string;
    initialData?: Omit<Client, 'birthday'> & { birthday?: string };
    redirect?: string;
    title?: string;
    showBackButton?: boolean;
    onSave?: (clientData: DocFilter) => void;
    onSaveAndCheck?: (clientData: DocFilter) => void;
}

/**
 * A form for creating or editing a client's document in the database.
 *
 * @param initialData Initial values to have in the input fields. If none are
 * given, empty strings will be used for text fields, 0 for number fields, and
 * the current date used for birthday
 * @param title Heading to show atop form
 * @param showBackButton Whether to show a "back to profile" button or not
 */
export default function ClientInfoForm({
    initialData = {} as Omit<Client, 'birthday'> & { birthday?: string },
    title = 'Client Form',
    onSave,
    onSaveAndCheck,
}: ClientInfoFormProps) {
    //TODO: Add Validation

    // Use default values if no initial data was passed in
    const defaultData = {
        id: initialData.id || '',
        firstName: initialData.firstName || '',
        lastName: initialData.lastName || '',
        firstNameLower: initialData.firstNameLower || '',
        lastNameLower: initialData.lastNameLower || '',
        middleInitial: initialData.middleInitial || '',
        birthday:
            initialData.birthday || new Date().toLocaleDateString('en-CA'),
        gender: initialData.gender || '',
        race: initialData.race || '',
        postalCode: initialData.postalCode || '',
        numKids: initialData.numKids?.toString() || '',
        notes: initialData.notes || '',
        isCheckedIn: !!initialData.isCheckedIn,
        isBanned: !!initialData.isBanned,
        unhoused: !!initialData.unhoused,
    };

    const [clientData, setClientData] = useState(defaultData);
    const handleSave = async () => {
        const data = {
            ...clientData,
            numKids: parseInt(clientData.numKids) || 0,
        };
        onSave && onSave(data);
    };

    const handleSaveAndCheck = async () => {
        const data = {
            ...clientData,
            numKids: parseInt(clientData.numKids) || 0,
        };
        onSaveAndCheck && onSaveAndCheck(data);
    };

    const handleChange = (key: any) => (e: any) => {
        let value = e.target.value;
        if (key === 'isBanned') value = e.target.checked;
        if (key === 'firstName')
            clientData.firstNameLower = value.toLowerCase();
        if (key === 'lastName') clientData.lastNameLower = value.toLowerCase();
        setClientData({ ...clientData, [key]: value });
    };
    const getFullName = () =>
        `${clientData.firstName} ${clientData.middleInitial} ${clientData.lastName}`;

    return (
        <div className={styles.container}>
            <div className={styles.titleContainer}>
                <h1>
                    {(!clientData.firstName &&
                        !clientData.middleInitial &&
                        !clientData.lastName &&
                        title) ||
                        getFullName()}
                </h1>
                <div className={styles.titleItem}>
                    <label htmlFor="unhoused">Unhoused</label>
                    <input
                        type="checkbox"
                        name="unhoused"
                        id="unhoused"
                        title="Check if client is unhoused"
                        defaultChecked={clientData.unhoused}
                        value={clientData.unhoused ? 'on' : 'off'}
                        onChange={handleChange('unhoused')}
                    />
                </div>
                <div className={styles.titleItem}>
                    <label htmlFor="isBanned">Ban</label>
                    <input
                        type="checkbox"
                        name="isBanned"
                        id="isBanned"
                        title="Check if client is banned"
                        defaultChecked={clientData.isBanned}
                        value={clientData.isBanned ? 'on' : 'off'}
                        onChange={handleChange('isBanned')}
                    />
                </div>
            </div>

            <form>
                <div className={styles.formRows}>
                    <div className={styles.formRowItem}>
                        <label htmlFor="firstName">First name</label>
                        <input
                            type="text"
                            value={clientData.firstName}
                            id="firstName"
                            onChange={handleChange('firstName')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="middleInitial">Middle initial</label>
                        <input
                            type="text"
                            value={clientData.middleInitial}
                            id="middleInitial"
                            onChange={handleChange('middleInitial')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="lastName">Last name</label>
                        <input
                            type="text"
                            value={clientData.lastName}
                            id="lastName"
                            onChange={handleChange('lastName')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="birthday">Birthday</label>
                        <input
                            type="date"
                            name="birthday"
                            id="birthday"
                            title="Enter client's birthday MM/DD/YYYY"
                            value={clientData.birthday}
                            onChange={handleChange('birthday')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="gender">Gender</label>
                        <input
                            type="text"
                            value={clientData.gender}
                            id="gender"
                            onChange={handleChange('gender')}
                        />
                        <br />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="race">Race</label>
                        <input
                            type="text"
                            value={clientData.race}
                            id="race"
                            onChange={handleChange('race')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="postalCode">Postal code</label>
                        <input
                            type="text"
                            value={clientData.postalCode}
                            id="postalCode"
                            onChange={handleChange('postalCode')}
                        />
                    </div>

                    <div className={styles.formRowItem}>
                        <label htmlFor="numKids">Number of Kids</label>
                        <input
                            type="number"
                            value={clientData.numKids}
                            id="numKids"
                            onChange={handleChange('numKids')}
                        />
                    </div>
                </div>

                <div>
                    <label htmlFor="notes">Notes</label>
                    <textarea
                        value={clientData.notes}
                        id="notes"
                        onChange={handleChange('notes')}
                        rows={5}
                    />
                </div>
            </form>

            <div className={styles.saveButtons}>
                {clientData.id && (
                    <>
                        <Link href={`/profile/${clientData.id}`}>
                            <button className={styles.backButton}>
                                Back to Profile
                            </button>
                        </Link>

                        <span />
                    </>
                )}

                <button className={styles.saveButton} onClick={handleSave}>
                    Save
                </button>

                <button
                    className={styles.saveButton}
                    onClick={handleSaveAndCheck}
                >
                    Save and check {clientData.isCheckedIn ? 'out' : 'in'}
                </button>
            </div>
        </div>
    );
}
