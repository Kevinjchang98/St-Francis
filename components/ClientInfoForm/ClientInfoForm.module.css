.container {
    display: flex;
    flex-direction: column;
}

.titleContainer {
    display: flex;
    flex-direction: row;
    align-items: center;
    border-bottom: 0.5px solid var(--gray-3);
}

.titleContainer h1 {
    flex-grow: 1;
}

.formRows {
    display: flex;
    flex-wrap: wrap;
}

.formRowItem {
    flex-grow: 1;
    flex-shrink: 0;
    flex-basis: calc(25% - 10px);
    margin: 5px;
    align-items: center;
    justify-content: center;
}

.container input,
.container label {
    display: block;
}

.container textarea {
    width: calc(100% - 10px);
    margin: 5px;
}
