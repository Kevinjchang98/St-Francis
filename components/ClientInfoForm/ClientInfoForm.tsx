import { addDoc, collection, doc, setDoc } from 'firebase/firestore';
import { useRouter } from 'next/navigation';
import { useState } from 'react';
import { firestore } from '../../firebase/firebase';

interface ClientInfoFormProps {
    id?: string;
    initialData?: {
        firstName: string;
        lastName: string;
        middleInitial: string;
        birthday: string; // TODO: Consider saving as timestamp
        gender: string;
        race: string;
        postalCode: string;
        numKids: number;
        notes: string;
        isCheckedIn: boolean;
        isBanned: boolean;
    };
}

/**
 * A form for creating or editing a client's document in the database.
 *
 * @param id The client's document we want to edit. If none is given, a new
 * client document will be created
 * @param initialData Initial values to have in the input fields. If none are
 * given, empty strings will be used for text fields, 0 for number fields, and
 * the current date used for birthday
 */
export default function ClientInfoForm({
    id = undefined,
    initialData = undefined,
}: ClientInfoFormProps) {
    // Use old info from props if it exists, else default values
    const [firstName, setFirstName] = useState<string>(
        initialData ? initialData.firstName : ''
    );
    const [lastName, setLastName] = useState<string>(
        initialData ? initialData.lastName : ''
    );
    const [middleInitial, setMiddleInitial] = useState<string>(
        initialData ? initialData.middleInitial : ''
    );
    const [birthday, setBirthday] = useState<any>(
        initialData
            ? initialData.birthday
            : new Date().toISOString().substring(0, 10)
    );
    const [gender, setGender] = useState<string>(
        initialData ? initialData.gender : ''
    );
    const [race, setRace] = useState<string>(
        initialData ? initialData.race : ''
    );
    const [postalCode, setPostalCode] = useState<string>(
        initialData ? initialData.postalCode : ''
    );
    const [numKids, setNumKids] = useState<number>(
        initialData ? initialData.numKids : 0
    );
    const [notes, setNotes] = useState<string>(
        initialData ? initialData.notes : ''
    );
    const [isCheckedIn, setIsCheckedIn] = useState<boolean>(
        initialData ? initialData.isCheckedIn : false
    );
    const [isBanned, setIsBanned] = useState<boolean>(
        initialData ? initialData.isBanned : false
    );
    const router = useRouter();

    // Creates a new doc with an automatically generated id for the client
    const addNewClient = async () => {
        // Only create if either first or last name is not empty
        if (firstName !== '' || lastName !== '') {
            await addDoc(collection(firestore, 'clients'), {
                firstName,
                lastName,
                middleInitial,
                birthday, // TODO: Consider saving as timestamp
                gender,
                race,
                postalCode,
                numKids,
                notes,
                isCheckedIn: false,
                isBanned,
            });
        }

        // Redirect back to index page
        router.push('/');
    };

    // Updates an existing client's document. If id is invalid will do nothing,
    // but this should never run with an invalid id
    const updateClientData = async () => {
        // Ensure id's not undefined
        if (id) {
            await setDoc(doc(firestore, 'clients', id), {
                firstName,
                lastName,
                middleInitial,
                birthday, // TODO: Consider saving as timestamp
                gender,
                race,
                postalCode,
                numKids,
                notes,
                isCheckedIn,
                isBanned,
            });

            router.push('/');
        }
    };

    return (
        <form
            onSubmit={(e) => {
                // Prevent redirect
                e.preventDefault();
                // If we have an id as a prop, update, else create new
                id ? updateClientData() : addNewClient();
            }}
        >
            <h2>Ban</h2>
            <input
                type="checkbox"
                name="isBanned"
                id="isBanned"
                defaultChecked={isBanned}
                value={isBanned ? 'on' : 'off'}
                onChange={(e) => {
                    setIsBanned(e.target.checked);
                }}
            />

            <h2>First name</h2>
            <input
                type="text"
                value={firstName}
                onChange={(e) => {
                    setFirstName(e.target.value);
                }}
            />
            <br />

            <h2>Middle initial</h2>
            <input
                type="text"
                value={middleInitial}
                onChange={(e) => {
                    setMiddleInitial(e.target.value);
                }}
            />
            <br />

            <h2>Last name</h2>
            <input
                type="text"
                value={lastName}
                onChange={(e) => {
                    setLastName(e.target.value);
                }}
            />
            <br />

            <h2>Birthday</h2>
            <input
                type="date"
                name="birthday"
                id="birthday"
                value={birthday}
                onChange={(e) => {
                    setBirthday(e.target.value);
                }}
            />

            <h2>Gender</h2>
            <input
                type="text"
                value={gender}
                onChange={(e) => {
                    setGender(e.target.value);
                }}
            />
            <br />

            <h2>Race</h2>
            <input
                type="text"
                value={race}
                onChange={(e) => {
                    setRace(e.target.value);
                }}
            />
            <br />

            <h2>Postal code</h2>
            <input
                type="text"
                value={postalCode}
                onChange={(e) => {
                    setPostalCode(e.target.value);
                }}
            />
            <br />

            <h2>Number of kids</h2>
            <input
                type="number"
                value={numKids}
                onChange={(e) => {
                    setNumKids(parseInt(e.target.value));
                }}
            />
            <br />

            <h2>Notes</h2>
            <textarea
                value={notes}
                onChange={(e) => {
                    setNotes(e.target.value);
                }}
                cols={40}
                rows={5}
            />
            <br />

            <button type="submit">Save</button>
        </form>
    );
}
