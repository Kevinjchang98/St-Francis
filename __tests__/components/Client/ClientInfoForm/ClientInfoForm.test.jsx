import ClientInfoForm from '@/components/Client/ClientInfoForm/ClientInfoForm';
import '@testing-library/jest-dom';
import { act, fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

describe('Client Info Form Component', () => {
    const mockClient = {
        firstName: 'First',
        lastName: 'Last',
        firstNameLower: 'first',
        lastNameLower: 'last',
        middleInitial: 'M',
        birthday: '2003-12-13',
        gender: 'Gender',
        race: 'Race',
        postalCode: '123456',
        numKids: 0,
        notes: 'Notes',
        isCheckedIn: false,
        isBanned: false,
        unhoused: false,
    };
    it('matches snapshot without any props', async () => {
        const tree = renderer.create(<ClientInfoForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly without initial data', async () => {
        await act(async () => {
            render(<ClientInfoForm />);
        });
        const title = screen.queryByRole('heading', {
            name: 'Client Form',
        });
        expect(title).toBeInTheDocument();
        expect(screen.queryByText('First name')).toBeInTheDocument();
        expect(screen.queryByText('Middle initial')).toBeInTheDocument();
        expect(screen.queryByText('Last name')).toBeInTheDocument();
        expect(screen.queryByText('Birthday')).toBeInTheDocument();
        expect(screen.queryByText('Gender')).toBeInTheDocument();
        expect(screen.queryByText('Race')).toBeInTheDocument();
        expect(screen.queryByText('Postal code')).toBeInTheDocument();
        expect(screen.queryByText('Number of Kids')).toBeInTheDocument();
        expect(screen.queryByText('Notes')).toBeInTheDocument();
        expect(screen.queryByText('Unhoused')).toBeInTheDocument();
        expect(screen.queryByText('Ban')).toBeInTheDocument();
    });

    it('renders correctly with initial client Data', async () => {
        await act(async () => {
            render(<ClientInfoForm initialData={mockClient} />);
        });
        const fullname = screen.queryByRole('heading', {
            name: `${mockClient.firstName} ${mockClient.middleInitial} ${mockClient.lastName}`,
        });
        expect(fullname).toBeInTheDocument();

        expect(screen.getByLabelText('Postal code').value).toBe(
            mockClient.postalCode
        );
        expect(screen.getByLabelText('First name').value).toBe(
            mockClient.firstName
        );
        expect(screen.getByLabelText('Middle initial').value).toBe(
            mockClient.middleInitial
        );
        expect(screen.getByLabelText('Last name').value).toBe(
            mockClient.lastName
        );
        expect(screen.getByLabelText('Birthday').value).toBe(
            mockClient.birthday
        );
        expect(screen.getByLabelText('Gender').value).toBe(mockClient.gender);
        expect(screen.getByLabelText('Race').value).toBe(mockClient.race);
        expect(screen.getByLabelText('Postal code').value).toBe(
            mockClient.postalCode
        );
        expect(screen.getByLabelText('Number of Kids').value).toBe(
            mockClient.numKids.toString()
        );
        expect(screen.getByLabelText('Notes').value).toBe(mockClient.notes);
        expect(screen.getByLabelText('Unhoused').value).toBe(
            (mockClient.unhoused && 'on') || 'off'
        );
        expect(screen.getByLabelText('Ban').value).toBe(
            (mockClient.isBanned && 'on') || 'off'
        );
    });

    it('saves correct output with mock onSubmit callback', async () => {
        // numKids will be convert to integer by the form on submit
        const mockOnSave = jest.fn((clientData) => clientData.numKids);
        render(
            <ClientInfoForm
                initialData={mockClient}
                actions={{ Save: mockOnSave }}
            />
        );
        const saveButton = screen.getByRole('button', { name: 'Save' });
        fireEvent.click(saveButton); // validate input fields before submit
        expect(mockOnSave).toHaveBeenCalledTimes(0);

        fireEvent.click(saveButton);
        expect(mockOnSave).toHaveBeenCalledTimes(1);
        expect(mockOnSave.mock.results[0].value).toBe(mockClient.numKids);
    });

    it('saves correct output with mock onSubmitAndCheck callback', async () => {
        const mockOnSaveAndCheck = jest.fn((clientData) => clientData.birthday);
        render(
            <ClientInfoForm
                initialData={{ ...mockClient, isCheckedIn: false }}
                actions={{
                    'Save and Check-in': mockOnSaveAndCheck,
                }}
            />
        );
        const saveButton = screen.getByRole('button', {
            name: 'Save and Check-in',
        });
        // Check input fields validation before submit
        fireEvent.click(saveButton);
        expect(mockOnSaveAndCheck).toHaveBeenCalledTimes(0);
        fireEvent.click(saveButton);
        expect(mockOnSaveAndCheck).toHaveBeenCalledTimes(1);
        expect(mockOnSaveAndCheck.mock.results[0].value).toBe(
            mockClient.birthday
        );
    });

    it('let user type in the firstName input fields correctly', async () => {
        render(<ClientInfoForm />);
        const firstNameInput = screen.getByLabelText('First name');
        expect(firstNameInput).toBeInTheDocument();
        await userEvent.type(firstNameInput, 'First');
        expect(firstNameInput.value).toBe('First');
    });
});
