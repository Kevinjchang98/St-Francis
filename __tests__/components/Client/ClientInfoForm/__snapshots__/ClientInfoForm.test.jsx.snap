// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Info Form Component matches snapshot without any props 1`] = `
<div
  className="container"
>
  <div
    className="titleContainer"
  >
    <h1>
      Client Form
    </h1>
    <div
      className="titleItem"
    >
      <label
        htmlFor="unhoused"
      >
        Unhoused
      </label>
      <input
        defaultChecked={false}
        id="unhoused"
        name="unhoused"
        onChange={[Function]}
        title="Check if client is unhoused"
        type="checkbox"
        value="off"
      />
    </div>
    <div
      className="titleItem"
    >
      <label
        htmlFor="isBanned"
      >
        Ban
      </label>
      <input
        defaultChecked={false}
        id="isBanned"
        name="isBanned"
        onChange={[Function]}
        title="Check if client is banned"
        type="checkbox"
        value="off"
      />
    </div>
  </div>
  <form>
    <div
      className="formRows"
    >
      <div
        className="formRowItem"
      >
        <label
          htmlFor="firstName"
        >
          First name
        </label>
        <input
          id="firstName"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="middleInitial"
        >
          Middle initial
        </label>
        <input
          id="middleInitial"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="lastName"
        >
          Last name
        </label>
        <input
          id="lastName"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="birthday"
        >
          Birthday
        </label>
        <input
          id="birthday"
          name="birthday"
          onChange={[Function]}
          title="Enter client's birthday MM/DD/YYYY"
          type="date"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="gender"
        >
          Gender
        </label>
        <input
          id="gender"
          onChange={[Function]}
          type="text"
          value=""
        />
        <br />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="race"
        >
          Race
        </label>
        <input
          id="race"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="postalCode"
        >
          Postal code
        </label>
        <input
          id="postalCode"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="formRowItem"
      >
        <label
          htmlFor="numKids"
        >
          Number of Kids
        </label>
        <input
          id="numKids"
          onChange={[Function]}
          type="number"
          value=""
        />
      </div>
    </div>
    <div>
      <label
        htmlFor="notes"
      >
        Notes
      </label>
      <textarea
        id="notes"
        onChange={[Function]}
        rows={5}
        value=""
      />
    </div>
  </form>
  <div
    className="saveButtons"
  >
    <button
      className="saveButton"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="saveButton"
      onClick={[Function]}
    >
      Save and check 
      in
    </button>
  </div>
</div>
`;
